crdsonnet/renderEngines/ast.libsonnet

```diff
@@ -0,0 +1,274 @@
+local j = import 'github.com/Duologic/jsonnet-libsonnet/main.libsonnet';
+local d = import 'github.com/jsonnet-libs/docsonnet/doc-util/main.libsonnet';
+
+{
+  local this = self,
+
+  nilvalue:: [],
+
+  // jsonnet-libsonnet provides some formatting with indentation and linebreaks, this customField removes that formatting, ideally this type of formatting should be configurable but I haven't found a good interface for that yet.
+  local customField = {
+    field(fieldname, expr, additive=false, hidden=false):
+      j.field.field(fieldname, expr, additive, hidden)
+      + {
+        toString(indent='', break=''):
+          j.field.field(fieldname, expr, additive, hidden).toString(),
+      },
+    func(fieldname, expr, params=[], hidden=false):
+      j.field.func(fieldname, expr, params, hidden)
+      + {
+        toString(indent='', break=''):
+          j.field.func(fieldname, expr, params, hidden).toString(),
+      },
+  },
+
+  nestInParents(name, parents, field)::
+    j.object.members([
+      std.foldr(
+        function(p, acc)
+          if p == name
+          then acc
+          else
+            customField.field(
+              j.fieldname.string(p),
+              j.object.members([
+                acc,
+              ]),
+              additive=true,
+            ),
+        parents,
+        field,
+      ),
+    ]),
+
+  functionName(name)::
+    local underscores = std.set(std.findSubstr('_', name));
+    local n = std.join('', [
+      if std.setMember(i - 1, underscores)
+      then std.asciiUpper(name[i])
+      else name[i]
+      for i in std.range(0, std.length(name) - 1)
+      if !std.setMember(i, underscores)
+    ]);
+    'with' + std.asciiUpper(n[0]) + n[1:],
+
+  functionHelp(functionName, schema)::
+    customField.field(
+      j.fieldname.string('#' + functionName),
+      j.literal(  // render docsonnet as literal to avoid docsonnet dependency
+        d.fn(
+          help=std.get(schema, 'description', ''),
+          args=(
+            if 'const' in schema
+            then []
+            else [
+              d.arg(
+                'value',
+                type=(
+                  if 'type' in schema
+                  then schema.type
+                  else 'string'
+                ),
+                default=(
+                  if 'default' in schema
+                  then schema.default
+                  else null
+                ),
+                enums=(
+                  if 'enum' in schema
+                  then schema.enum
+                  else null
+                )
+              ),
+            ]
+          )
+        ),
+      )
+    ),
```

Comment:
These could have some nice unittests!

Reply:
Yeah, I would like to unit test jsonnet-libsonnet too, would increase confidence when using it here then.